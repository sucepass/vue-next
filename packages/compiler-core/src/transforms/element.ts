import { NodeTransform, TransformContext } from '../transform'
import {
  NodeTypes,
  ElementTypes,
  CallExpression,
  ObjectExpression,
  ElementNode,
  DirectiveNode,
  ExpressionNode,
  ArrayExpression,
  createCallExpression,
  createArrayExpression,
  createObjectProperty,
  createExpression,
  createObjectExpression
} from '../ast'
import { isArray } from '@vue/shared'
import { createCompilerError, ErrorCodes } from '../errors'
import {
  CREATE_VNODE,
  APPLY_DIRECTIVES,
  RESOLVE_DIRECTIVE,
  RESOLVE_COMPONENT
} from '../runtimeConstants'

const toValidId = (str: string): string => str.replace(/[^\w]/g, '')

// generate a JavaScript AST for this element's codegen
export const prepareElementForCodegen: NodeTransform = (node, context) => {
  if (node.type === NodeTypes.ELEMENT) {
    if (
      node.tagType === ElementTypes.ELEMENT ||
      node.tagType === ElementTypes.COMPONENT
    ) {
      const isComponent = node.tagType === ElementTypes.COMPONENT
      const hasProps = node.props.length > 0
      const hasChildren = node.children.length > 0
      let runtimeDirectives: DirectiveNode[] | undefined
      let componentIdentifier: string | undefined

      if (isComponent) {
        context.imports.add(RESOLVE_COMPONENT)
        componentIdentifier = `_component_${toValidId(node.tag)}`
        context.statements.push(
          `const ${componentIdentifier} = ${RESOLVE_COMPONENT}(${JSON.stringify(
            node.tag
          )})`
        )
      }

      const args: CallExpression['arguments'] = [
        isComponent ? componentIdentifier! : `"${node.tag}"`
      ]
      // props
      if (hasProps) {
        const { props, directives } = buildProps(node, context)
        args.push(props)
        runtimeDirectives = directives
      }
      // children
      if (hasChildren) {
        if (!hasProps) {
          // placeholder for null props, but use `0` for more condense code
          args.push(`0`)
        }
        args.push(isComponent ? buildSlots(node, context) : node.children)
      }

      const { loc } = node
      context.imports.add(CREATE_VNODE)
      const vnode = createCallExpression(CREATE_VNODE, args, loc)

      if (runtimeDirectives && runtimeDirectives.length) {
        context.imports.add(APPLY_DIRECTIVES)
        node.codegenNode = createCallExpression(
          APPLY_DIRECTIVES,
          [
            vnode,
            createArrayExpression(
              runtimeDirectives.map(dir => {
                return createDirectiveArgs(dir, context)
              }),
              loc
            )
          ],
          loc
        )
      } else {
        node.codegenNode = vnode
      }
    } else if (node.tagType === ElementTypes.SLOT) {
      // <slot [name="xxx"]/>
      // TODO
    } else if (node.tagType === ElementTypes.TEMPLATE) {
      // do nothing
    }
  }
}

function buildProps(
  { loc, props }: ElementNode,
  context: TransformContext
): {
  props: ObjectExpression | CallExpression
  directives: DirectiveNode[]
} {
  let properties: ObjectExpression['properties'] = []
  const mergeArgs: Array<ObjectExpression | ExpressionNode> = []
  const runtimeDirectives: DirectiveNode[] = []

  for (let i = 0; i < props.length; i++) {
    // static attribute
    const prop = props[i]
    if (prop.type === NodeTypes.ATTRIBUTE) {
      const { loc, name, value } = prop
      properties.push(
        createObjectProperty(
          createExpression(name, true, loc),
          createExpression(
            value ? value.content : '',
            true,
            value ? value.loc : loc
          ),
          loc
        )
      )
    } else {
      // directives
      // special case for v-bind with no argument
      if (prop.name === 'bind' && !prop.arg) {
        if (prop.exp) {
          if (properties.length) {
            mergeArgs.push(createObjectExpression(properties, loc))
            properties = []
          }
          mergeArgs.push(prop.exp)
        } else {
          context.onError(
            createCompilerError(ErrorCodes.X_V_BIND_NO_EXPRESSION, prop.loc)
          )
        }
        continue
      }

      const directiveTransform = context.directiveTransforms[prop.name]
      if (directiveTransform) {
        const { props, needRuntime } = directiveTransform(prop, context)
        if (isArray(props)) {
          properties.push(...props)
        } else {
          properties.push(props)
        }
        if (needRuntime) {
          runtimeDirectives.push(prop)
        }
      } else {
        // no built-in transform, this is a user custom directive.
        runtimeDirectives.push(prop)
      }
    }
  }

  let ret: ObjectExpression | CallExpression

  // has v-bind="object", wrap with mergeProps
  if (mergeArgs.length) {
    if (properties.length) {
      mergeArgs.push(createObjectExpression(properties, loc))
    }
    if (mergeArgs.length > 1) {
      ret = createCallExpression(`mergeProps`, mergeArgs, loc)
    } else {
      // single v-bind with nothing else - no need for a mergeProps call
      ret = createObjectExpression(properties, loc)
    }
  } else {
    ret = createObjectExpression(properties, loc)
  }

  return {
    props: ret,
    directives: runtimeDirectives
  }
}

function createDirectiveArgs(
  dir: DirectiveNode,
  context: TransformContext
): ArrayExpression {
  // inject import for `resolveDirective`
  context.imports.add(RESOLVE_DIRECTIVE)
  // inject statement for resolving directive
  const dirIdentifier = `_directive_${toValidId(dir.name)}`
  context.statements.push(
    `const ${dirIdentifier} = ${RESOLVE_DIRECTIVE}(${JSON.stringify(dir.name)})`
  )
  const dirArgs: ArrayExpression['elements'] = [dirIdentifier]
  const { loc } = dir
  if (dir.exp) dirArgs.push(dir.exp)
  if (dir.arg) dirArgs.push(dir.arg)
  if (Object.keys(dir.modifiers).length) {
    dirArgs.push(
      createObjectExpression(
        dir.modifiers.map(modifier =>
          createObjectProperty(
            createExpression(modifier, true, loc),
            createExpression(`true`, false, loc),
            loc
          )
        ),
        loc
      )
    )
  }
  return createArrayExpression(dirArgs, dir.loc)
}

function buildSlots(
  { loc, children }: ElementNode,
  context: TransformContext
): ObjectExpression {
  const slots = createObjectExpression([], loc)
  // TODO

  return slots
}
