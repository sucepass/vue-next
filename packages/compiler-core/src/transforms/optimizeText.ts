import { NodeTransform } from '../transform'
import {
  NodeTypes,
  TemplateChildNode,
  TextNode,
  InterpolationNode,
  CompoundExpressionNode
} from '../ast'

const isText = (
  node: TemplateChildNode
): node is TextNode | InterpolationNode =>
  node.type === NodeTypes.INTERPOLATION || node.type === NodeTypes.TEXT

// Merge adjacent text nodes and expressions into a single expression
// e.g. <div>abc {{ d }} {{ e }}</div> should have a single expression node as child.
export const optimizeText: NodeTransform = node => {
  if (node.type === NodeTypes.ROOT || node.type === NodeTypes.ELEMENT) {
    // perform the transform on node exit so that all expressions have already
    // been processed.
    return () => {
      const children = node.children
      let currentContainer: CompoundExpressionNode | undefined = undefined
      for (let i = 0; i < children.length; i++) {
        const child = children[i]
        if (isText(child)) {
          for (let j = i + 1; j < children.length; j++) {
            const next = children[j]
            if (isText(next)) {
              if (!currentContainer) {
                currentContainer = children[i] = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  loc: child.loc,
                  children: [child]
                }
              }
              // merge adjacent text node into current
              currentContainer.children.push(` + `, next)
              children.splice(j, 1)
              j--
            } else {
              currentContainer = undefined
              break
            }
          }
        }
      }
    }
  }
}
