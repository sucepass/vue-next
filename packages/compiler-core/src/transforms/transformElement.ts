import { NodeTransform, TransformContext } from '../transform'
import {
  NodeTypes,
  ElementTypes,
  CallExpression,
  ObjectExpression,
  ElementNode,
  DirectiveNode,
  ExpressionNode,
  ArrayExpression,
  createCallExpression,
  createArrayExpression,
  createObjectProperty,
  createSimpleExpression,
  createObjectExpression,
  Property
} from '../ast'
import { PatchFlags, PatchFlagNames, isSymbol } from '@vue/shared'
import { createCompilerError, ErrorCodes } from '../errors'
import {
  CREATE_VNODE,
  APPLY_DIRECTIVES,
  RESOLVE_DIRECTIVE,
  RESOLVE_COMPONENT,
  MERGE_PROPS,
  TO_HANDLERS
} from '../runtimeHelpers'
import { getInnerRange, isVSlot, toValidAssetId } from '../utils'
import { buildSlots } from './vSlot'

// some directive transforms (e.g. v-model) may return a symbol for runtime
// import, which should be used instead of a resolveDirective call.
const directiveImportMap = new WeakMap<DirectiveNode, symbol>()

// generate a JavaScript AST for this element's codegen
export const transformElement: NodeTransform = (node, context) => {
  if (node.type === NodeTypes.ELEMENT) {
    if (
      node.tagType === ElementTypes.ELEMENT ||
      node.tagType === ElementTypes.COMPONENT ||
      // <template> with v-if or v-for are ignored during traversal.
      // <template> without v-slot should be treated as a normal element.
      (node.tagType === ElementTypes.TEMPLATE && !node.props.some(isVSlot))
    ) {
      // perform the work on exit, after all child expressions have been
      // processed and merged.
      return () => {
        const isComponent = node.tagType === ElementTypes.COMPONENT
        let hasProps = node.props.length > 0
        let patchFlag: number = 0
        let runtimeDirectives: DirectiveNode[] | undefined
        let dynamicPropNames: string[] | undefined

        if (isComponent) {
          context.helper(RESOLVE_COMPONENT)
          context.components.add(node.tag)
        }

        const args: CallExpression['arguments'] = [
          isComponent ? toValidAssetId(node.tag, `component`) : `"${node.tag}"`
        ]
        // props
        if (hasProps) {
          const propsBuildResult = buildProps(node, context)
          patchFlag = propsBuildResult.patchFlag
          dynamicPropNames = propsBuildResult.dynamicPropNames
          runtimeDirectives = propsBuildResult.directives
          if (!propsBuildResult.props) {
            hasProps = false
          } else {
            args.push(propsBuildResult.props)
          }
        }
        // children
        const hasChildren = node.children.length > 0
        if (hasChildren) {
          if (!hasProps) {
            args.push(`null`)
          }
          if (isComponent) {
            const { slots, hasDynamicSlots } = buildSlots(node, context)
            args.push(slots)
            if (hasDynamicSlots) {
              patchFlag |= PatchFlags.DYNAMIC_SLOTS
            }
          } else if (node.children.length === 1) {
            const child = node.children[0]
            const type = child.type
            const hasDynamicTextChild =
              type === NodeTypes.INTERPOLATION ||
              type === NodeTypes.COMPOUND_EXPRESSION
            if (hasDynamicTextChild) {
              patchFlag |= PatchFlags.TEXT
            }
            // pass directly if the only child is a text node
            // (plain / interpolation / expression)
            if (hasDynamicTextChild || type === NodeTypes.TEXT) {
              args.push(child)
            } else {
              args.push(node.children)
            }
          } else {
            args.push(node.children)
          }
        }
        // patchFlag & dynamicPropNames
        if (patchFlag !== 0) {
          if (!hasChildren) {
            if (!hasProps) {
              args.push(`null`)
            }
            args.push(`null`)
          }
          if (__DEV__) {
            const flagNames = Object.keys(PatchFlagNames)
              .map(Number)
              .filter(n => n > 0 && patchFlag & n)
              .map(n => PatchFlagNames[n])
              .join(`, `)
            args.push(patchFlag + ` /* ${flagNames} */`)
          } else {
            args.push(patchFlag + '')
          }
          if (dynamicPropNames && dynamicPropNames.length) {
            args.push(
              `[${dynamicPropNames.map(n => JSON.stringify(n)).join(`, `)}]`
            )
          }
        }

        const { loc } = node
        const vnode = createCallExpression(
          context.helper(CREATE_VNODE),
          args,
          loc
        )

        if (runtimeDirectives && runtimeDirectives.length) {
          node.codegenNode = createCallExpression(
            context.helper(APPLY_DIRECTIVES),
            [
              vnode,
              createArrayExpression(
                runtimeDirectives.map(dir => buildDirectiveArgs(dir, context)),
                loc
              )
            ],
            loc
          )
        } else {
          node.codegenNode = vnode
        }
      }
    }
  }
}

export type PropsExpression = ObjectExpression | CallExpression | ExpressionNode

export function buildProps(
  node: ElementNode,
  context: TransformContext,
  props: ElementNode['props'] = node.props
): {
  props: PropsExpression | undefined
  directives: DirectiveNode[]
  patchFlag: number
  dynamicPropNames: string[]
} {
  const elementLoc = node.loc
  const isComponent = node.tagType === ElementTypes.COMPONENT
  let properties: ObjectExpression['properties'] = []
  const mergeArgs: PropsExpression[] = []
  const runtimeDirectives: DirectiveNode[] = []

  // patchFlag analysis
  let patchFlag = 0
  let hasRef = false
  let hasClassBinding = false
  let hasStyleBinding = false
  let hasDynamicKeys = false
  const dynamicPropNames: string[] = []

  const analyzePatchFlag = ({ key, value }: Property) => {
    if (key.type === NodeTypes.SIMPLE_EXPRESSION && key.isStatic) {
      if (value.type !== NodeTypes.SIMPLE_EXPRESSION || !value.isStatic) {
        const name = key.content
        if (name === 'ref') {
          hasRef = true
        } else if (name === 'class') {
          hasClassBinding = true
        } else if (name === 'style') {
          hasStyleBinding = true
        } else if (name !== 'key') {
          dynamicPropNames.push(name)
        }
      }
    } else {
      hasDynamicKeys = true
    }
  }

  for (let i = 0; i < props.length; i++) {
    // static attribute
    const prop = props[i]
    if (prop.type === NodeTypes.ATTRIBUTE) {
      const { loc, name, value } = prop
      if (name === 'ref') {
        hasRef = true
      }
      properties.push(
        createObjectProperty(
          createSimpleExpression(
            name,
            true,
            getInnerRange(loc, 0, name.length)
          ),
          createSimpleExpression(
            value ? value.content : '',
            true,
            value ? value.loc : loc
          )
        )
      )
    } else {
      // directives
      const { name, arg, exp, loc } = prop

      // skip v-slot - it is handled by its dedicated transform.
      if (name === 'slot') {
        if (!isComponent) {
          context.onError(
            createCompilerError(ErrorCodes.X_V_SLOT_MISPLACED, loc)
          )
        }
        continue
      }

      // special case for v-bind and v-on with no argument
      const isBind = name === 'bind'
      const isOn = name === 'on'
      if (!arg && (isBind || isOn)) {
        hasDynamicKeys = true
        if (exp) {
          if (properties.length) {
            mergeArgs.push(
              createObjectExpression(dedupeProperties(properties), elementLoc)
            )
            properties = []
          }
          if (isBind) {
            mergeArgs.push(exp)
          } else {
            // v-on="obj" -> toHandlers(obj)
            mergeArgs.push({
              type: NodeTypes.JS_CALL_EXPRESSION,
              loc,
              callee: context.helper(TO_HANDLERS),
              arguments: [exp]
            })
          }
        } else {
          context.onError(
            createCompilerError(
              isBind
                ? ErrorCodes.X_V_BIND_NO_EXPRESSION
                : ErrorCodes.X_V_ON_NO_EXPRESSION,
              loc
            )
          )
        }
        continue
      }

      const directiveTransform = context.directiveTransforms[name]
      if (directiveTransform) {
        // has built-in directive transform.
        const { props, needRuntime } = directiveTransform(prop, node, context)
        props.forEach(analyzePatchFlag)
        properties.push(...props)
        if (needRuntime) {
          runtimeDirectives.push(prop)
          if (isSymbol(needRuntime)) {
            directiveImportMap.set(prop, needRuntime)
          }
        }
      } else {
        // no built-in transform, this is a user custom directive.
        runtimeDirectives.push(prop)
      }
    }
  }

  let propsExpression: PropsExpression | undefined = undefined

  // has v-bind="object" or v-on="object", wrap with mergeProps
  if (mergeArgs.length) {
    if (properties.length) {
      mergeArgs.push(
        createObjectExpression(dedupeProperties(properties), elementLoc)
      )
    }
    if (mergeArgs.length > 1) {
      propsExpression = createCallExpression(
        context.helper(MERGE_PROPS),
        mergeArgs,
        elementLoc
      )
    } else {
      // single v-bind with nothing else - no need for a mergeProps call
      propsExpression = mergeArgs[0]
    }
  } else if (properties.length) {
    propsExpression = createObjectExpression(
      dedupeProperties(properties),
      elementLoc
    )
  }

  // patchFlag analysis
  if (hasDynamicKeys) {
    patchFlag |= PatchFlags.FULL_PROPS
  } else {
    if (hasClassBinding) {
      patchFlag |= PatchFlags.CLASS
    }
    if (hasStyleBinding) {
      patchFlag |= PatchFlags.STYLE
    }
    if (dynamicPropNames.length) {
      patchFlag |= PatchFlags.PROPS
    }
  }
  if (patchFlag === 0 && (hasRef || runtimeDirectives.length > 0)) {
    patchFlag |= PatchFlags.NEED_PATCH
  }

  return {
    props: propsExpression,
    directives: runtimeDirectives,
    patchFlag,
    dynamicPropNames
  }
}

// Dedupe props in an object literal.
// Literal duplicated attributes would have been warned during the parse phase,
// however, it's possible to encounter duplicated `onXXX` handlers with different
// modifiers. We also need to merge static and dynamic class / style attributes.
// - onXXX handlers / style: merge into array
// - class: merge into single expression with concatenation
function dedupeProperties(properties: Property[]): Property[] {
  const knownProps: Record<string, Property> = {}
  const deduped: Property[] = []
  for (let i = 0; i < properties.length; i++) {
    const prop = properties[i]
    // dynamic keys are always allowed
    if (prop.key.type === NodeTypes.COMPOUND_EXPRESSION || !prop.key.isStatic) {
      deduped.push(prop)
      continue
    }
    const name = prop.key.content
    const existing = knownProps[name]
    if (existing) {
      if (
        name === 'style' ||
        name === 'class' ||
        name.startsWith('on') ||
        name.startsWith('vnode')
      ) {
        mergeAsArray(existing, prop)
      }
      // unexpected duplicate, should have emitted error during parse
    } else {
      knownProps[name] = prop
      deduped.push(prop)
    }
  }
  return deduped
}

function mergeAsArray(existing: Property, incoming: Property) {
  if (existing.value.type === NodeTypes.JS_ARRAY_EXPRESSION) {
    existing.value.elements.push(incoming.value)
  } else {
    existing.value = createArrayExpression(
      [existing.value, incoming.value],
      existing.loc
    )
  }
}

function buildDirectiveArgs(
  dir: DirectiveNode,
  context: TransformContext
): ArrayExpression {
  const dirArgs: ArrayExpression['elements'] = []
  const runtime = directiveImportMap.get(dir)
  if (runtime) {
    context.helper(runtime)
    dirArgs.push(context.helperString(runtime))
  } else {
    // inject statement for resolving directive
    context.helper(RESOLVE_DIRECTIVE)
    context.directives.add(dir.name)
    dirArgs.push(toValidAssetId(dir.name, `directive`))
  }
  const { loc } = dir
  if (dir.exp) dirArgs.push(dir.exp)
  if (dir.arg) {
    if (!dir.exp) {
      dirArgs.push(`void 0`)
    }
    dirArgs.push(dir.arg)
  }
  if (Object.keys(dir.modifiers).length) {
    if (!dir.arg) {
      if (!dir.exp) {
        dirArgs.push(`void 0`)
      }
      dirArgs.push(`void 0`)
    }
    dirArgs.push(
      createObjectExpression(
        dir.modifiers.map(modifier =>
          createObjectProperty(
            modifier,
            createSimpleExpression(`true`, false, loc)
          )
        ),
        loc
      )
    )
  }
  return createArrayExpression(dirArgs, dir.loc)
}
