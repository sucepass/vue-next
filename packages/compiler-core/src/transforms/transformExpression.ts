// - Parse expressions in templates into compound expressions so that each
//   identifier gets more accurate source-map locations.
//
// - Prefix identifiers with `_ctx.` so that they are accessed from the render
//   context
//
// - This transform is only applied in non-browser builds because it relies on
//   an additional JavaScript parser. In the browser, there is no source-map
//   support and the code is wrapped in `with (this) { ... }`.

import { parseScript } from 'meriyah'
import { walk } from 'estree-walker'
import { NodeTransform, TransformContext } from '../transform'
import { NodeTypes, createExpression, ExpressionNode } from '../ast'
import { Node, Function, Identifier, Property } from 'estree'
import { advancePositionWithClone } from '../utils'
export const transformExpression: NodeTransform = (node, context) => {
  if (node.type === NodeTypes.EXPRESSION && !node.isStatic) {
    processExpression(node, context)
  } else if (node.type === NodeTypes.ELEMENT) {
    // handle directives on element
    for (let i = 0; i < node.props.length; i++) {
      const prop = node.props[i]
      if (prop.type === NodeTypes.DIRECTIVE) {
        if (prop.exp) {
          processExpression(prop.exp, context)
        }
        if (prop.arg && !prop.arg.isStatic) {
          if (prop.name === 'class') {
            // TODO special expression optimization for classes
            processExpression(prop.arg, context)
          } else {
            processExpression(prop.arg, context)
          }
        }
      }
    }
  }
}

const simpleIdRE = /^[a-zA-Z$_][\w$]*$/

const isFunction = (node: Node): node is Function =>
  /Function(Expression|Declaration)$/.test(node.type)

const isPropertyKey = (node: Node, parent: Node) =>
  parent.type === 'Property' && parent.key === node && !parent.computed

// cache node requires
let _parseScript: typeof parseScript
let _walk: typeof walk

interface PrefixMeta {
  prefix?: string
  start: number
  end: number
}

// Important: since this function uses Node.js only dependencies, it should
// always be used with a leading !__BROWSER__ check so that it can be
// tree-shaken from the browser build.
export function processExpression(
  node: ExpressionNode,
  context: TransformContext
) {
  if (!context.prefixIdentifiers) {
    return
  }
  // lazy require dependencies so that they don't end up in rollup's dep graph
  // and thus can be tree-shaken in browser builds.
  const parseScript =
    _parseScript || (_parseScript = require('meriyah').parseScript)
  const walk = _walk || (_walk = require('estree-walker').walk)

  // fast path if expression is a simple identifier.
  if (simpleIdRE.test(node.content)) {
    if (!context.identifiers[node.content]) {
      node.content = `_ctx.${node.content}`
    }
    return
  }

  let ast
  try {
    ast = parseScript(`(${node.content})`, { ranges: true }) as any
  } catch (e) {
    context.onError(e)
    return
  }

  const ids: (Identifier & PrefixMeta)[] = []
  const knownIds = Object.create(context.identifiers)

  // walk the AST and look for identifiers that need to be prefixed with `_ctx.`.
  walk(ast, {
    enter(node: Node & PrefixMeta, parent) {
      if (node.type === 'Identifier') {
        if (!ids.includes(node)) {
          if (!knownIds[node.name] && shouldPrefix(node, parent)) {
            if (
              isPropertyKey(node, parent) &&
              (parent as Property).value === node
            ) {
              // property shorthand like { foo }, we need to add the key since we
              // rewrite the value
              node.prefix = `${node.name}: `
            }
            node.name = `_ctx.${node.name}`
            ids.push(node)
          } else if (!isPropertyKey(node, parent)) {
            // also generate sub-expressioms for other identifiers for better
            // source map support. (except for property keys which are static)
            ids.push(node)
          }
        }
      } else if (isFunction(node)) {
        // walk function expressions and add its arguments to known identifiers
        // so that we don't prefix them
        node.params.forEach(p =>
          walk(p, {
            enter(child) {
              if (child.type === 'Identifier') {
                knownIds[child.name] = true
                ;(
                  (node as any)._scopeIds ||
                  ((node as any)._scopeIds = new Set())
                ).add(child.name)
              }
            }
          })
        )
      }
    },
    leave(node: any) {
      if (node._scopeIds) {
        node._scopeIds.forEach((id: string) => {
          delete knownIds[id]
        })
      }
    }
  })

  // We break up the coumpound expression into an array of strings and sub
  // expressions (for identifiers that have been prefixed). In codegen, if
  // an ExpressionNode has the `.children` property, it will be used instead of
  // `.content`.
  const full = node.content
  const children: ExpressionNode['children'] = []
  ids.sort((a, b) => a.start - b.start)
  ids.forEach((id, i) => {
    const last = ids[i - 1] as any
    const leadingText = full.slice(last ? last.end - 1 : 0, id.start - 1)
    if (leadingText.length || id.prefix) {
      children.push(leadingText + (id.prefix || ``))
    }
    const source = full.slice(id.start - 1, id.end - 1)
    children.push(
      createExpression(id.name, false, {
        source,
        start: advancePositionWithClone(node.loc.start, source, id.start + 2),
        end: advancePositionWithClone(node.loc.start, source, id.end + 2)
      })
    )
    if (i === ids.length - 1 && id.end - 1 < full.length) {
      children.push(full.slice(id.end - 1))
    }
  })

  if (children.length) {
    // mark it empty so that it's more noticeable in case another transform or
    // codegen forget to handle `.children` first.
    node.content = __DEV__ ? `[[REMOVED]]` : ``
    node.children = children
  }
}

const globals = new Set(
  (
    'Infinity,undefined,NaN,isFinite,isNaN,' +
    'parseFloat,parseInt,decodeURI,decodeURIComponent,encodeURI,encodeURIComponent,' +
    'Math,Number,Date,Array,Object,Boolean,String,RegExp,Map,Set,JSON,Intl,' +
    'require,' + // for webpack
    'arguments,'
  ) // parsed as identifier but is a special keyword...
    .split(',')
)

function shouldPrefix(identifier: Identifier, parent: Node) {
  if (
    !(
      isFunction(parent) &&
      // not id of a FunctionDeclaration
      ((parent as any).id === identifier ||
        // not a params of a function
        parent.params.includes(identifier))
    ) &&
    // not a key of Property
    !(
      isPropertyKey(identifier, parent) &&
      // shorthand keys should be prefixed
      !((parent as Property).value === identifier)
    ) &&
    // not a property of a MemberExpression
    !(
      parent.type === 'MemberExpression' &&
      parent.property === identifier &&
      !parent.computed
    ) &&
    // not in an Array destructure pattern
    !(parent.type === 'ArrayPattern') &&
    // skip globals + commonly used shorthands
    !globals.has(identifier.name)
  ) {
    return true
  }
}
