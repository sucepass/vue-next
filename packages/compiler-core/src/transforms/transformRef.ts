import { NodeTransform } from '../transform'
import {
  NodeTypes,
  ElementTypes,
  createSimpleExpression,
  createCompoundExpression
} from '../ast'
import { findProp } from '../utils'

// Convert ref="foo" to `:ref="[_ctx, 'foo']"` so that the ref contains the
// correct owner instance even inside slots.
export const transformRef: NodeTransform = node => {
  if (
    !(
      node.type === NodeTypes.ELEMENT &&
      (node.tagType === ElementTypes.ELEMENT ||
        node.tagType === ElementTypes.COMPONENT)
    )
  ) {
    return
  }
  const ref = findProp(node, 'ref')
  if (!ref) return
  const refKey =
    ref.type === NodeTypes.ATTRIBUTE
      ? ref.value
        ? createSimpleExpression(ref.value.content, true, ref.value.loc)
        : null
      : ref.exp
  if (refKey) {
    node.props[node.props.indexOf(ref)] = {
      type: NodeTypes.DIRECTIVE,
      name: `bind`,
      arg: createSimpleExpression(`ref`, true, ref.loc),
      exp: createCompoundExpression([`[_ctx, `, refKey, `]`]),
      modifiers: [],
      loc: ref.loc
    }
  }
}
