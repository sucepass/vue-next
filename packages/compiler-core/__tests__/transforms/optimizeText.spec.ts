import {
  CompilerOptions,
  parse,
  transform,
  NodeTypes,
  generate
} from '../../src'
import { optimizeText } from '../../src/transforms/optimizeText'
import { transformExpression } from '../../src/transforms/transformExpression'
import { transformElement } from '../../src/transforms/transformElement'

function transformWithTextOpt(template: string, options: CompilerOptions = {}) {
  const ast = parse(template)
  transform(ast, {
    nodeTransforms: [
      ...(options.prefixIdentifiers ? [transformExpression] : []),
      optimizeText,
      transformElement
    ],
    ...options
  })
  return ast
}

describe('compiler: optimize interpolation', () => {
  test('no consecutive text', () => {
    const root = transformWithTextOpt(`{{ foo }}`)
    expect(root.children[0]).toMatchObject({
      type: NodeTypes.INTERPOLATION,
      content: {
        content: `foo`
      }
    })
    expect(generate(root).code).toMatchSnapshot()
  })

  test('consecutive text', () => {
    const root = transformWithTextOpt(`{{ foo }} bar {{ baz }}`)
    expect(root.children.length).toBe(1)
    expect(root.children[0]).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        { type: NodeTypes.INTERPOLATION, content: { content: `foo` } },
        ` + `,
        { type: NodeTypes.TEXT, content: ` bar ` },
        ` + `,
        { type: NodeTypes.INTERPOLATION, content: { content: `baz` } }
      ]
    })
    expect(generate(root).code).toMatchSnapshot()
  })

  test('consecutive text between elements', () => {
    const root = transformWithTextOpt(`<div/>{{ foo }} bar {{ baz }}<div/>`)
    expect(root.children.length).toBe(3)
    expect(root.children[0].type).toBe(NodeTypes.ELEMENT)
    expect(root.children[1]).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        { type: NodeTypes.INTERPOLATION, content: { content: `foo` } },
        ` + `,
        { type: NodeTypes.TEXT, content: ` bar ` },
        ` + `,
        { type: NodeTypes.INTERPOLATION, content: { content: `baz` } }
      ]
    })
    expect(root.children[2].type).toBe(NodeTypes.ELEMENT)
    expect(generate(root).code).toMatchSnapshot()
  })

  test('consecutive text mixed with elements', () => {
    const root = transformWithTextOpt(
      `<div/>{{ foo }} bar {{ baz }}<div/>{{ foo }} bar {{ baz }}<div/>`
    )
    expect(root.children.length).toBe(5)
    expect(root.children[0].type).toBe(NodeTypes.ELEMENT)
    expect(root.children[1]).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        { type: NodeTypes.INTERPOLATION, content: { content: `foo` } },
        ` + `,
        { type: NodeTypes.TEXT, content: ` bar ` },
        ` + `,
        { type: NodeTypes.INTERPOLATION, content: { content: `baz` } }
      ]
    })
    expect(root.children[2].type).toBe(NodeTypes.ELEMENT)
    expect(root.children[3]).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        { type: NodeTypes.INTERPOLATION, content: { content: `foo` } },
        ` + `,
        { type: NodeTypes.TEXT, content: ` bar ` },
        ` + `,
        { type: NodeTypes.INTERPOLATION, content: { content: `baz` } }
      ]
    })
    expect(root.children[4].type).toBe(NodeTypes.ELEMENT)
    expect(generate(root).code).toMatchSnapshot()
  })

  test('with prefixIdentifiers: true', () => {
    const root = transformWithTextOpt(`{{ foo }} bar {{ baz + qux }}`, {
      prefixIdentifiers: true
    })
    expect(root.children.length).toBe(1)
    expect(root.children[0]).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        { type: NodeTypes.INTERPOLATION, content: { content: `_ctx.foo` } },
        ` + `,
        { type: NodeTypes.TEXT, content: ` bar ` },
        ` + `,
        {
          type: NodeTypes.INTERPOLATION,
          content: {
            type: NodeTypes.COMPOUND_EXPRESSION,
            children: [{ content: `_ctx.baz` }, ` + `, { content: `_ctx.qux` }]
          }
        }
      ]
    })
    expect(
      generate(root, {
        prefixIdentifiers: true
      }).code
    ).toMatchSnapshot()
  })
})
