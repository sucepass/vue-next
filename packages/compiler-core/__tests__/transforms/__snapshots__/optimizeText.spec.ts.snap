// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: optimize interpolation consecutive text 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { toString: _toString } = _Vue
    
    return _toString(foo) + \\" bar \\" + _toString(baz)
  }
}"
`;

exports[`compiler: optimize interpolation consecutive text between elements 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { createVNode: _createVNode, toString: _toString, createBlock: _createBlock, Fragment: _Fragment, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(_Fragment, null, [
      _createVNode(\\"div\\"),
      _toString(foo) + \\" bar \\" + _toString(baz),
      _createVNode(\\"div\\")
    ]))
  }
}"
`;

exports[`compiler: optimize interpolation consecutive text mixed with elements 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { createVNode: _createVNode, toString: _toString, createBlock: _createBlock, Fragment: _Fragment, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(_Fragment, null, [
      _createVNode(\\"div\\"),
      _toString(foo) + \\" bar \\" + _toString(baz),
      _createVNode(\\"div\\"),
      _toString(foo) + \\" bar \\" + _toString(baz),
      _createVNode(\\"div\\")
    ]))
  }
}"
`;

exports[`compiler: optimize interpolation no consecutive text 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { toString: _toString } = _Vue
    
    return _toString(foo)
  }
}"
`;

exports[`compiler: optimize interpolation with prefixIdentifiers: true 1`] = `
"const { toString } = Vue

return function render() {
  const _ctx = this
  return toString(_ctx.foo) + \\" bar \\" + toString(_ctx.baz + _ctx.qux)
}"
`;
